File: haskell/intro/Intro.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE ScopedTypeVariables #-}

import Control.Monad
import Control.Monad.ST
import Control.Monad.Trans.State
import Control.Monad.IO.Class
import Data.Text (Text)
import Data.Word
import Foreign.C.Types
import SDL hiding (get)
import qualified SDL.Font as TTF

type Game = StateT GameState IO

data GameState = GameState
  { window :: Window
  , renderer :: Renderer
  , timer :: Word32
  , bitmap :: Surface
  , statestack :: [Game ()]
  }

windowWidth = 800
windowHeight = 600
windowCaption = "SDL Introduction"

fps = 30
frameRate = 1000 / fps

main :: IO ()
main = do
  initializeAll
  window <- createWindow windowCaption defaultWindow
  let dr = defaultRenderer { rendererType = SoftwareRenderer }
  renderer <- createRenderer window (-1) dr
  timer <- ticks
  bitmap <- loadBMP "../../shared/background.bmp"
  runStateT runGame (GameState window renderer timer bitmap [menu])
  return ()

-- Functions without explicit signatures are :: Game ()
runGame :: Game ()
runGame = do
  initGame
  gameLoop
  shutdown

gameLoop = do
  stack <- gets statestack 
  unless (null stack) $ do
    head stack
    gameLoop

initGame = do
  SDL.initialize [InitVideo, InitTimer]
  TTF.initialize

shutdown = do
  b <- gets bitmap
  freeSurface b
  quit
  TTF.quit

menu = do
  fhe <- frameHasElapsed
  when fhe $ do
    handleInput handleMenuInput
    clearScreen
    displayText "(S)tart game" 350 250 20 255 255 255 0 0 0
    displayText "(Q)uit game" 350 275 20 255 255 255 0 0 0
    finishFrame

game = do
  fhe <- frameHasElapsed
  when fhe $ do
    handleInput handleGameInput
    clearScreen
    drawBackground
    finishFrame

exit = do
  fhe <- frameHasElapsed
  when fhe $ do
    handleInput handleExitInput
    clearScreen
    displayText "Quit Game (Y or N)?" 350 250 12 255 255 255 0 0 0
    finishFrame

drawBackground = do
  renderer <- gets renderer
  bitmap <- gets bitmap
  texture <- createTextureFromSurface renderer bitmap
  copy renderer texture Nothing Nothing

displayText :: Text -> CInt -> CInt -> CInt -> Word8 -> Word8 -> Word8 -> 
               Word8 -> Word8 -> Word8 -> Game ()
displayText text x y size_ fR fG fB bR bG bB = do
  renderer <- gets renderer
  font <- TTF.load "../../shared/arial.ttf" 24
  surface <- TTF.shaded font (V4 fR fG fB 1) (V4 bR bG bB 1) text 
  (w', h') <- TTF.size font text
  let (w, h) :: (CInt, CInt) = (fromIntegral w', fromIntegral h')
  texture <- createTextureFromSurface renderer surface
  copy renderer texture Nothing $ Just $ Rectangle (P (V2 x y)) (V2 w h)
  freeSurface surface
  TTF.free font

clearScreen = do
  renderer <- gets renderer
  clear renderer

frameHasElapsed :: Game Bool
frameHasElapsed = do
  t <- gets timer
  t' <- ticks 
  return $ fromIntegral t' - fromIntegral t >= frameRate

finishFrame = do
  window <- gets window
  renderer <- gets renderer
  present renderer
  updateWindowSurface window
  state <- get
  t <- ticks
  put state { timer = t }

handleMenuInput :: KeyboardEventData -> Game ()
handleMenuInput e =
  case keysymKeycode (keyboardEventKeysym e) of
    KeycodeQ -> pushState exit
    KeycodeS -> pushState game
    _ -> return ()

handleGameInput :: KeyboardEventData -> Game ()
handleGameInput e =
  case keysymKeycode (keyboardEventKeysym e) of
    KeycodeEscape -> popState
    _ -> return ()
            
handleExitInput :: KeyboardEventData -> Game ()
handleExitInput e =
  case keysymKeycode (keyboardEventKeysym e) of
    KeycodeY -> clearStateStack
    KeycodeN -> popState
    _ -> return ()

handleInput :: (KeyboardEventData -> Game ()) -> Game ()
handleInput f = do
  event <- pollEvent
  case event of 
    Nothing -> return ()
    Just event ->
      case eventPayload event of
        QuitEvent -> clearStateStack
        KeyboardEvent k -> f k
        _ -> return ()

clearStateStack = do
  state <- get
  put state { statestack = [] }
  
pushState s = do
  state <- get
  stack <- gets statestack
  put state { statestack = s : stack }

popState = do
  state <- get
  stack <- gets statestack
  put state { statestack = tail stack }

File: c++/intro/Intro.cpp
#include <SDL2/SDL.h>
#include <SDL2/SDL_TTF.h>

#include <stack>
#include <iostream>

using namespace std;

const int WINDOW_WIDTH = 800;
const int WINDOW_HEIGHT = 600;
const string WINDOW_CAPTION = "SDL Introduction";

const int FRAMES_PER_SECOND = 30;
const double FRAME_RATE = 1000/FRAMES_PER_SECOND;

struct StateStruct {
  void (*StatePointer)();
};

stack<StateStruct> g_StateStack;

SDL_Surface* g_Bitmap = nullptr;
SDL_Window* g_Window = nullptr;
SDL_Renderer* g_Renderer = nullptr;
SDL_Event g_Event;
int g_Timer;

void Menu();
void Game();
void Exit();

void Init();
void Shutdown();
void DrawBackground();
void ClearScreen();
void DisplayText(string text, int x, int y, int size, int fR, int fG, int fB,
                 int bR, int bG, int bB);
void HandleMenuInput();
void HandleGameInput();
void HandleExitInput();
bool FrameHasElapsed();
void FinishFrame();
void ClearStateStack();

int main(int arg, char**argv) {
  Init();

  while (!g_StateStack.empty()) {
    g_StateStack.top().StatePointer();
  }
  Shutdown();

  return 0;
}

void Init() {
  SDL_Init(SDL_INIT_VIDEO | SDL_INIT_TIMER);

  g_Window = SDL_CreateWindow(WINDOW_CAPTION.c_str(), SDL_WINDOWPOS_UNDEFINED, SDL_WINDOWPOS_UNDEFINED, WINDOW_WIDTH, WINDOW_HEIGHT, SDL_WINDOW_OPENGL);
  g_Renderer = SDL_CreateRenderer(g_Window, -1, 0);
  g_Timer = SDL_GetTicks();
  g_Bitmap = SDL_LoadBMP("../../shared/background.bmp");

  StateStruct state;
  state.StatePointer = Exit;
  g_StateStack.push(state);

  state.StatePointer = Menu;
  g_StateStack.push(state);

  TTF_Init();
}

void Shutdown() {
  SDL_FreeSurface(g_Bitmap);
  SDL_Quit();
  TTF_Quit();
}

void Menu() {
  if (!FrameHasElapsed()) return;

  HandleMenuInput();
  ClearScreen();
  DisplayText("(S)tart Game", 350, 250, 20, 255, 255, 255, 0, 0, 0);
  DisplayText("(Q)uit Game", 350, 275, 20, 255, 255, 255, 0, 0, 0);
  FinishFrame();
}

void Game() {
  if (!FrameHasElapsed()) return;

  HandleGameInput();
  ClearScreen();
  DrawBackground();
  FinishFrame();
}

void Exit() {
  if (!FrameHasElapsed()) return;

  HandleExitInput();
  ClearScreen();
  DisplayText("Quit Game (Y or N)?", 350, 250, 12, 255, 255, 255, 0, 0, 0);
  FinishFrame();
}

void DrawBackground() {
  SDL_Rect source = {0, 0, WINDOW_WIDTH, WINDOW_HEIGHT};
  SDL_Rect destination = {0, 0, WINDOW_WIDTH, WINDOW_HEIGHT};

  SDL_Texture* texture = SDL_CreateTextureFromSurface(g_Renderer, g_Bitmap);
  SDL_RenderCopy(g_Renderer, texture, nullptr, nullptr);
}

void ClearScreen() {
  SDL_RenderClear(g_Renderer);
  SDL_UpdateWindowSurface(g_Window);
}

void DisplayText(string text, int x, int y, int size, int fR, int fG, int fB,
                 int bR, int bG, int bB) {
  TTF_Font* font = TTF_OpenFont("../../shared/arial.ttf", 24);

  SDL_Color foreground = {static_cast<Uint8>(fR), static_cast<Uint8>(fG), static_cast<Uint8>(fB)};
  SDL_Color background = {static_cast<Uint8>(bR), static_cast<Uint8>(bG), static_cast<Uint8>(bB)};

  SDL_Surface* temp = TTF_RenderText_Shaded(font, text.c_str(), foreground, background);

  int w, h;
  TTF_SizeText(font, text.c_str(), &w, &h);
  SDL_Rect destination = {x, y, w, h};

  SDL_Texture* texture = SDL_CreateTextureFromSurface(g_Renderer, temp);
  SDL_RenderCopy(g_Renderer, texture, nullptr, &destination);

  SDL_FreeSurface(temp);
  TTF_CloseFont(font);
}

void HandleMenuInput() {
  if (!SDL_PollEvent(&g_Event)) return;

  if (g_Event.type == SDL_QUIT) {
    ClearStateStack();
  }
  else if (g_Event.type == SDL_KEYDOWN) {
    switch (g_Event.key.keysym.sym) {
      case SDLK_ESCAPE:
      case SDLK_q:
        g_StateStack.pop();
        break;

      case SDLK_s:
        StateStruct temp;
        temp.StatePointer = Game;
        g_StateStack.push(temp);
        break;

      default:
        break;
    }
  }
}

void HandleGameInput() {
  if (!SDL_PollEvent(&g_Event)) return;

  if (g_Event.type == SDL_QUIT) {
    ClearStateStack();
  }
  else if (g_Event.type == SDL_KEYDOWN) {
    switch(g_Event.key.keysym.sym) {
      case SDLK_ESCAPE:
        g_StateStack.pop();
        break;

      default:
        break;
    }
  }
}

void HandleExitInput() {
  if (!SDL_PollEvent(&g_Event)) return;

  if (g_Event.type == SDL_QUIT) {
    ClearStateStack();
  }
  else if (g_Event.type == SDL_KEYDOWN) {
    switch(g_Event.key.keysym.sym) {
      case SDLK_y:
        g_StateStack.pop();
        break;

      case SDLK_n:
        StateStruct temp;
        temp.StatePointer = Menu;
        g_StateStack.push(temp);
        break;

      default:
        break;
    }
  }
}

bool FrameHasElapsed() {
  return ((SDL_GetTicks() - g_Timer) >= FRAME_RATE);
}

void FinishFrame() {
    SDL_UpdateWindowSurface(g_Window);
    SDL_RenderPresent(g_Renderer);
    g_Timer = SDL_GetTicks();
}

void ClearStateStack() {
  while (!g_StateStack.empty()) {
    g_StateStack.pop();
  }
}
